src/psyclone/psyir/transformations/intrinsics/mms_base_trans.py

```diff
@@ -0,0 +1,326 @@
+        elif (isinstance(dimension_ref, Reference) and
+              dimension_ref.symbol.is_constant):
+            dimension_literal = dimension_ref.symbol.constant_value
+        # else exception is handled by the validate method.
+
+        # Determine the dimension and extent of the array
+        ndims = None
+        if len(array_ref.children) == 0:
+            # Note, the potential 'if not array_ref.symbol.is_array:'
+            # exception is already handled by the validate method.
+            ndims = len(array_ref.symbol.shape)
+
+            loop_bounds = []
+            for idx, shape in enumerate(array_ref.symbol.shape):
+                if shape in [ArrayType.Extent.DEFERRED,
+                             ArrayType.Extent.ATTRIBUTE]:
+                    # runtime extent using LBOUND and UBOUND required
+                    lbound = BinaryOperation.create(
+                        BinaryOperation.Operator.LBOUND,
+                        Reference(array_ref.symbol),
+                        Literal(str(idx+1), INTEGER_TYPE))
+                    ubound = BinaryOperation.create(
+                        BinaryOperation.Operator.UBOUND,
+                        Reference(array_ref.symbol),
+                        Literal(str(idx+1), INTEGER_TYPE))
+                    loop_bounds.append((lbound, ubound))
+                elif isinstance(shape, ArrayType.ArrayBounds):
+                    # array extent is defined in the array declaration
+                    loop_bounds.append(shape)
+                #  Note, the validate method guarantees that an else
+                #  clause is not required.
+        else:
+            # The validate method guarantees that this is an array
+            # reference.
+            loop_bounds = []
+            ndims = len(array_ref.children)
+            for shape in array_ref.children:
+                loop_bounds.append((shape.start.copy(), shape.stop.copy()))
+
+        # Determine the datatype of the array's values and create a
+        # scalar of that type
+        array_intrinsic = array_ref.symbol.datatype.intrinsic
+        array_precision = array_ref.symbol.datatype.precision
+        scalar_type = ScalarType(array_intrinsic, array_precision)
+
+        symbol_table = node.scope.symbol_table
+        assignment = node.ancestor(Assignment)
+
+        datatype = scalar_type
+        array_reduction = False
+        if dimension_ref and ndims > 1:
+            array_reduction = True
+            # We are reducing from one array to another
+            shape = []
+            for idx, bounds in enumerate(loop_bounds):
+                if int(dimension_literal.value)-1 == idx:
+                    pass
+                else:
+                    shape.append(bounds)
+            datatype = ArrayType(scalar_type, shape)
+
+        array_ref = node.children[0].detach()
+
+        # Create temporary variable based on the name of the intrinsic.
+        var_symbol = symbol_table.new_symbol(
```

Comment:
I think we could hit trouble here if we don't know the extent(s) of the array that we are performing the reduction into. e.g. the array argument to MAXVAL might only just have been allocated.

Reply:
Ooh, you are on fire. Yes, you are right, we end up declaring an array with ubound and lbound for the original array before the original array is allocated. The newly allocated array should also be made allocatable in this case and allocated/deallocated just before and after the modified code.

Reply:
OK, I think that is now working. I didn't bother deallocating as the array will go out of scope at the end of the routine anyway. There is a new test called `test_allocate_dim` in `mms_base_trans_test.py`.